### DISCUSSION TITLE
navigation.replaceAll() reverted animation

### DISCUSSION THREAD

**Discussion by Tboy70** [arkivanov/Decompose#677] [Q&A] [answered]
Hi everyone, 

I'm trying to use the Decompose library for navigation. Here is my need : 

- I have some screens flow like this : A --> B --> C
- After C, comes D screen, but this screen should become **the only one**

I managed to do this using the `replaceAll()` function of the library. However, the slide animation direction (right to left) that was correct from A to C is reverted when D comes. Is there a way to make it keep the same right to left direction ? 

Thank you very much for your help !

**Comment by arkivanov** [comment]
I recall this was fixed in version 2.2.0: https://github.com/arkivanov/Decompose/releases/tag/2.2.0

What version are you using?

**Reply by Tboy70** [reply]
I'm using the 2.2.2 version (last stable).
Maybe I'm making something wrong. Here is basically my code : 

```
@OptIn(ExperimentalDecomposeApi::class)
    @Suppress("LongMethod")
    private fun createChild(
        configuration: Configuration,
        componentContext: ComponentContext
    ): Child {
        return when (configuration) {
             ...
            Configuration.Payment -> Child.Payment(
                PaymentComponent(
                    componentContext = componentContext,
                    onNavigateToSaleSummary = {
                        navigation.replaceAll(Configuration.SaleSummary)
                    },
                    onBackPressed = navigation::pop
                )
            )

            Configuration.SaleSummary -> Child.SaleSummary(
                SaleSummaryComponent(
                    componentContext = componentContext,
                )
            )
    }
}
```

**Reply by arkivanov** [reply]
This looks good. It would be also nice to see how you specify the animation.

I have just checked the sample app after applying the following patch, the animation is playing forward on `replaceAll`. You can checkout the sample, apply the patch, push a few Counter components to the stack, then click "Prev" - `replaceAll` replaces the entire stack with one *new* component and the animation looks correct.

```diff
Index: sample/shared/shared/src/commonMain/kotlin/com/arkivanov/sample/shared/counters/DefaultCountersComponent.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/sample/shared/shared/src/commonMain/kotlin/com/arkivanov/sample/shared/counters/DefaultCountersComponent.kt b/sample/shared/shared/src/commonMain/kotlin/com/arkivanov/sample/shared/counters/DefaultCountersComponent.kt
--- a/sample/shared/shared/src/commonMain/kotlin/com/arkivanov/sample/shared/counters/DefaultCountersComponent.kt	(revision 477818583868b238fa8d4cb59eedff68d57c1447)
+++ b/sample/shared/shared/src/commonMain/kotlin/com/arkivanov/sample/shared/counters/DefaultCountersComponent.kt	(date 1710596405021)
@@ -7,6 +7,7 @@
 import com.arkivanov.decompose.router.stack.pop
 import com.arkivanov.decompose.router.stack.popTo
 import com.arkivanov.decompose.router.stack.push
+import com.arkivanov.decompose.router.stack.replaceAll
 import com.arkivanov.decompose.value.Value
 import com.arkivanov.sample.shared.counters.counter.CounterComponent
 import com.arkivanov.sample.shared.counters.counter.DefaultCounterComponent
@@ -37,7 +38,14 @@
             title = "Counter ${config.index}",
             isBackEnabled = config.isBackEnabled,
             onNext = { navigation.push(Config(index = config.index + 1, isBackEnabled = true)) },
-            onPrev = navigation::pop,
+            onPrev = {
+                navigation.replaceAll(
+                    Config(
+                        index = _childStack.value.items.maxOf { it.configuration.index } + 1,
+                        isBackEnabled = false,
+                    )
+                )
+            },
         )
 
     override fun onBackClicked() {
```

**Reply by arkivanov** [reply]
Please also make sure that `Configuration.SaleSummary` is not present in the stack when you call `replaceAll`.

**Reply by Tboy70** [reply]
Thanks for your answer. 
The animation is specified like this : 
`Children(
            stack = childStack,
            animation = stackAnimation(slide())
        ) { child ->
        ...`
        
Yes, Configuration.SaleSummary is not present in the stack. 

I will test the patch, thanks again.

**Reply by arkivanov** [reply]
Let me know if you need any help.

**Reply by Tboy70** [reply]
OK, so I have tested the sample. You're right, the slide animation is on the correct direction, even with the replaceAll function. 
I must be mistaken somewhere, just need to find where

**Reply by arkivanov** [reply]
Thanks for the update!

> I'm using the 2.2.2 version

Do you also use version `2.2.0` for Compose extensions (`compose-extensions-compose` or `compose-extensions-jetbrains`)?

**Reply by Tboy70** [reply]
That was that, I didn't have those dependencies ! Thank you, it's working on Android now :) 

I still have the issue on iOS however. Certainly also a dependencies issue

**Reply by Tboy70** [reply]
Well it's not finish after all. Some import are unknown now 🤔
![Capture d’écran 2024-03-16 à 17 38 39](https://github.com/arkivanov/Decompose/assets/7996019/c769505d-1b1f-448b-a3a9-8f5638c298a7)

**Reply by arkivanov** [reply]
Could you please mention what versions you are using now? And what are the imports?

**Reply by Tboy70** [reply]
Sure. I forgot to precise but I'm using CMM (Compose Multiplatform). I don't know if changes something compared with KMM.

Here is what I have right now. With that, the gradle sync work, but the imports are not resolved

![Capture d’écran 2024-03-16 à 18 07 13](https://github.com/arkivanov/Decompose/assets/7996019/1d5b9a28-0a07-45d8-b48c-d7f5b3d64e95)

```
sourceSets {
        androidMain.dependencies {
            ...
            implementation("com.arkivanov.decompose:decompose:2.2.2")
            implementation("com.arkivanov.decompose:extensions-compose-jetbrains:2.2.2")
            ...
        }
        commonMain.dependencies {
            ...
            implementation("com.arkivanov.decompose:decompose:2.2.2")
            ...
        }
        iosMain.dependencies {
            ...
            implementation("com.arkivanov.decompose:decompose:2.2.2-compose-experimental")
            implementation("com.arkivanov.decompose:extensions-compose-jetbrains:2.2.2-compose-experimental")
             ...
        }
    }
```

**Reply by Tboy70** [reply]
I think I got it --> com.arkivanov.decompose:extensions-compose-jetbrains:2.2.0 instead of 2.2.2, put in androidMain AND commonMain seem to work

**Reply by arkivanov** [reply, marked as answer]
Thanks! I believe you defined Decompose dependencies incorrectly. Could you please try in the following way?

```
commonMain.dependencies {
    implementation("com.arkivanov.decompose:decompose:2.2.2-compose-experimental")
    implementation("com.arkivanov.decompose:extensions-compose-jetbrains:2.2.2-compose-experimental")
}
```

And you don't need to repeat those dependencies in other source sets like `androidMain` and `iosMain`. Also, it's better to use the exact same version for all Decompose modules, in this case it's `2.2.2-compose-experimental`.

**Reply by Tboy70** [reply]
Thank you it seems to work now. Very easy at the end 🙂
